### ISSUE TITLE
[API] We should review how we use OSGi Semantic Versioning

### ISSUE THREAD

**Issue by cdealti** [eclipse/kura#226]
For the Kura API we use OSGi Semantic Versioning of packages.
Every exported package is qualified with a version that we upgrade if something changes.

Since we aim to backward compatibility we only make changes to the API that require a change in the minor number of the exported package.

For example in #194 we are adding methods to `org.eclipse.kura.configuration.ConfigurationService`.
Due to this we also have to change the minor version to the  `org.eclipse.kura.configuration` package exported in the MANIFEST of the `org.eclipse.kura.api` bundle, from version 1.0.0 to 1.1.0.

This will not cause problems to **consumers** of the `org.eclipse.kura.configuration` API if they are properly importing the `org.eclipse.kura.configuration` package:

```
Import-Package: org.eclipse.kura.configuration;version="[1.0,2.0)",
```

However, consumers of `org.eclipse.kura.configuration` are rare.

Virtually all Kura apps are **implementors** of the  `org.eclipse.kura.configuration` API for example because they implement `org.eclipse.kura.configuration.ConfigurableComponent` and if they properly use semantic versioning:

```
Import-Package: org.eclipse.kura.configuration;version="[1.0,1.1)",
```

They will have their imports not satisfied which is good for Semantic Versioning but it's not good for customers wanting to upgrade Kura without upgrading the apps (and in this example there would be no need to upgrade the apps).

There is really nothing we can do in this case, only hope that customer apps are not following Semantic Versioning in this case.

The situation is the same for the various listeners we have in the API.

For the future:
We should have a finer granularity for packages to have better control on versioning: `ConfigurableComponent` should belong to an `org.eclipse.kura.configuration.component`

**Comment by LucaDazi** [comment]
Agree on putting marker interfaces, abstract classes and listeners on dedicated packages.

We could start duplicating classes where needed and deprecating the old ones.
We'll have to take that step, sooner or later.

**Comment by cdealti** [comment]
Something has been done for the Data Service and Data Transport Service listeners.
New ``org.eclipse.kura.data.listener`` and `` org.eclipse.kura.data.transport.listener`` have been created to properly support the multiple connections feature.
The old [listeners](https://github.com/eclipse/kura/blob/develop/kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/data/DataServiceListener.java) in ``org.eclipse.kura.data`` have been deprecated.

**Comment by cdealti** [comment]
To clearly state the requirements:

1. Apps should only import packages exported by the ``org.eclipse.kura.bundle``. Ideally Kura implementation bundles should not export anything and if they have they should contain the word **internal** in their names (for new designs). An app should not import internal packages because they can be frequently subject to incompatible changes.

2. Apps should import the Kura API packages following the Semantic Versioning

3. The Kura API must be backward compatible. This means that the packages contained in ``org.eclipse.kura.api`` cannot change their **major** version.

4. Clearly in the Kura API there is an implicit assumption on whether an API package is meant to be consumed or implemented by apps. The rule (for new designs) is that if an API is meant to be implemented by apps it will go in a separate package. Good candidates are the above listeners. Unfortunately, the Kura API is not structured like this. Gradually, we will improve things by deprecating interfaces and classes which should belong to an "meant to be implemented by apps" packages.

5. An API package meant to be implemented by apps must be backward compatible, meaning that it cannot even change its **minor** number. The reason is obvious. An app importing the package and following Semantic Versioning would be broken by such a change and we don't want our customers to complain they have to recompile and redeploy their apps. Unfortunately this is beyond Semantic Versioning capabilities.

6. An API package meant to be consumed by apps is typically provided by Kura. The version of these packages can change in the **minor** number. There might be cases where an alternative implementation of an API is provided externally to Kura. For example an alternative implementation of the org.eclipse.kura.cloud package. These implementations, being providers, would be broken by the above change and need to be updated.

In other words and unless planned well in advance:

- It's not allowed to change the **major** version in any of the packages of the Kura API bundle.
- It's not allowed to change the **minor** version in any of the packages of the Kura API bundle meant to be implemented by apps.
- On a change of the **major** API number all deprecated APIs must be removed.

See also #352.

**Comment by cdealti** [comment]
Wiki page created [here](https://github.com/eclipse/kura/wiki/Kura-Semantic-Versioning)
